### PR TITLE
[monotouch-test] Make it work with Mac Catalyst.

### PR DESCRIPTION
**Pull request by rolfbjarne**
* Fix system version checks to work properly on Mac Catalyst (which uses the macOS
  version as its system version).

* Add the framework-specific defines to the build for monotouch-test.csproj (using
  the generated response files), this way we can use them in the tests.

* Sprinkle conditionals in numerous places - I tried using either framework-specific
  or XAMCORE_3_0 whenever that made since (instead of using Mac Catalyst as a condition).

* Updated a few tests to use non-deprecated API (because the deprecated API often
  doesn't exist in Mac Catalyst).

* Also a few minor API fixes to make any corresponding tests compile.

monotouch-test isn't quite green yet - the remaining tests will be fixed in
later pull requests.

### CODE REVIEW

**Comment by spouliot** [comment]
So the version returned is the macOS one, like `11.0`
but all API are decorated with versions like `[MacCatalyst (14,0)]` ?
😞

**Review by spouliot — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Comment by rolfbjarne** [comment]
> So the version returned is the macOS one, like `11.0`
> but all API are decorated with versions like `[MacCatalyst (14,0)]` ?
> 😞

I had a deeper look, and it seems the `@available` block in Objective-C reads the `iOSSupportVersion` from `/System/Library/CoreServices/SystemVersion.plist`: 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>ProductBuildVersion</key>
	<string>19H114</string>
	<key>ProductCopyright</key>
	<string>1983-2020 Apple Inc.</string>
	<key>ProductName</key>
	<string>Mac OS X</string>
	<key>ProductUserVisibleVersion</key>
	<string>10.15.7</string>
	<key>ProductVersion</key>
	<string>10.15.7</string>
	<key>iOSSupportVersion</key>
	<string>13.6</string>
</dict>
</plist>
```

So I guess this is something we might want to discuss.

**Review by tj-devel709 — APPROVED** [review]

File: src/ObjCRuntime/Runtime.cs [hunk 1]
```diff
@@ -1806,6 +1806,21 @@ internal static IntPtr AllocGCHandle (object value)
...
 		}
+
+#if __MACCATALYST__
+		static string _iOSSupportVersion;
+		internal static string iOSSupportVersion {
+			get {
+				if (_iOSSupportVersion == null) {
+					// This is how Apple does it: https://github.com/llvm/llvm-project/blob/62ec4ac90738a5f2d209ed28c822223e58aaaeb7/lldb/source/Host/macosx/objcxx/HostInfoMacOSX.mm#L100-L105
+					using (var dict = NSMutableDictionary.FromFile ("/System/Library/CoreServices/SystemVersion.plist")) {
+						_iOSSupportVersion = dict.ObjectForKey ((NSString) "iOSSupportVersion").ToString ();
```
Comment (spouliot) [inline src/ObjCRuntime/Runtime.cs:1817]: that `NSString` could be collected before being used
https://github.com/xamarin/xamarin-macios/issues/10146

**Review by spouliot — APPROVED** [review]

**Review by mandel-macaque — APPROVED** [review]

File: src/ObjCRuntime/Runtime.cs (hunk 1 above)
Comment (rolfbjarne) [reply src/ObjCRuntime/Runtime.cs]: Fixed.

File: tests/monotouch-test/AVFoundation/CaptureDeviceTest.cs [hunk 2]
```diff
@@ -34,7 +34,12 @@ public void CompareConstantEnum ()
 				Compare (AVMediaType.MetadataObject, AVMediaTypes.MetadataObject);
 
 			// obsoleted in iOS 6, removed in iOS12
-			if (TestRuntime.CheckSystemVersion (PlatformName.iOS, 12,0))
+			var removed = TestRuntime.CheckSystemVersion (PlatformName.iOS, 12, 0);
+#if __MACCATALYST__
+			removed = true;
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply tests/monotouch-test/AVFoundation/CaptureDeviceTest.cs]: Fixed.

**Review by spouliot — APPROVED** [review]

**Review by mandel-macaque — APPROVED** [review]

**Comment by rolfbjarne** [comment]
Test failure is unrelated (https://github.com/xamarin/maccore/issues/2378)

**Pull request merged**